hal/src/main/native/athena/SerialPort.cpp

```diff
@@ -1,174 +1,492 @@
+
+  serialPort->tty.c_lflag &= ~(ICANON | ECHO | ISIG);
+  serialPort->tty.c_iflag &= ~(IXON | IXOFF | IXANY);
+  /* Raw output mode, sends the raw and unprocessed data  (send as it is).
+   * If it is in canonical mode and sending new line char then CR
+   * will be added as prefix and send as CR LF
+   */
+  serialPort->tty.c_oflag = ~OPOST;
+
+  tcflush(serialPort->portId, TCIOFLUSH);
+  if (tcsetattr(serialPort->portId, TCSANOW, &serialPort->tty) != 0) {
+    *status = errno;
+    close(serialPort->portId);
+    serialPortHandles->Free(handle);
+    return HAL_kInvalidHandle;
   }
+  return handle;
+}
 
-  *status = viOpen(resourceManagerHandle, const_cast<char*>(portName.c_str()),
-                   VI_NULL, VI_NULL,
-                   reinterpret_cast<ViSession*>(&portHandles[port]));
-  if (*status > 0) *status = 0;
+void HAL_CloseSerial(HAL_SerialPortHandle handle, int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  serialPortHandles->Free(handle);
+
+  if (port) {
+    close(port->portId);
+  }
 }
 
-void HAL_InitializeSerialPortDirect(HAL_SerialPort port, const char* portName,
-                                    int32_t* status) {
-  *status = viOpen(resourceManagerHandle, const_cast<char*>(portName), VI_NULL,
-                   VI_NULL, reinterpret_cast<ViSession*>(&portHandles[port]));
-  if (*status > 0) *status = 0;
+int HAL_GetSerialFD(HAL_SerialPortHandle handle, int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return -1;
+  }
+  return port->portId;
 }
 
-void HAL_SetSerialBaudRate(HAL_SerialPort port, int32_t baud, int32_t* status) {
-  *status = viSetAttribute(portHandles[port], VI_ATTR_ASRL_BAUD, baud);
-  if (*status > 0) *status = 0;
+#define BAUDCASE(BAUD)        \
+  case BAUD:                  \
+    port->baudRate = B##BAUD; \
+    break;
+
+void HAL_SetSerialBaudRate(HAL_SerialPortHandle handle, int32_t baud,
+                           int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+
+  switch (baud) {
+    BAUDCASE(50)
+    BAUDCASE(75)
+    BAUDCASE(110)
+    BAUDCASE(134)
+    BAUDCASE(150)
+    BAUDCASE(200)
+    BAUDCASE(300)
+    BAUDCASE(600)
+    BAUDCASE(1200)
+    BAUDCASE(1800)
+    BAUDCASE(2400)
+    BAUDCASE(4800)
+    BAUDCASE(9600)
+    BAUDCASE(19200)
+    BAUDCASE(38400)
+    BAUDCASE(57600)
+    BAUDCASE(115200)
+    BAUDCASE(230400)
+    BAUDCASE(460800)
+    BAUDCASE(500000)
+    BAUDCASE(576000)
+    BAUDCASE(921600)
+    BAUDCASE(1000000)
+    BAUDCASE(1152000)
+    BAUDCASE(1500000)
+    BAUDCASE(2000000)
+    BAUDCASE(2500000)
+    BAUDCASE(3000000)
+    BAUDCASE(3500000)
+    BAUDCASE(4000000)
+    default:
+      *status = PARAMETER_OUT_OF_RANGE;
+      return;
+  }
+  cfsetospeed(&port->tty, static_cast<speed_t>(port->baudRate));
+  cfsetispeed(&port->tty, static_cast<speed_t>(port->baudRate));
```

Comment:
Should we check errors returned from cfsetospeed and cfsetispeed as well?

Reply:
Maybe. It would message if the speed isn't valid, so yeah probably a good idea.